### ISSUE TITLE
Mixins and interfaces order in the declarations phase

### ISSUE THREAD

**Issue by scheglov** [dart-lang/language#3437]
For this library
```dart
@AddFunction('f3')
mixin class X3 {}

@AddFunction('f2')
mixin class X2 {}

@AddFunction('f1')
class X1 with X2, X3 {}
```
Do we follow the right to left rule, and generate
```dart
library augment 'test.dart';

void f3() {}
void f2() {}
void f1() {}
```

or the source order, and generate
```dart
library augment 'test.dart';

void f2() {}
void f3() {}
void f1() {}
```

@jakemac53

**Comment by jakemac53** [comment]
It would just be source order here - the right to left rule is only for when multiple macros are applied to the exact same declaration.

**Note**: In phase 2 there is a special rule "Macros are applied to superclasses, mixins, and interfaces first, in Phase 2". That doesn't actually change things here, but if you moved the class `X1` above the mixin declarations, it wouldn't change the order of the macros (it would still be `f3`, `f2`, `f1`, because all macros on the mixing would run first). This ensures that you can see all the generated members of your super interfaces.

**Comment by scheglov** [comment]
Yes, this is what I'm working now in the ordering - apply to superclasses, mixins, and interfaces first.

Oh, and the related question - the order is exactly this - first superclasses, then mixins in the source order, then interfaces in the source order, right?

Recursively, so if there are not  yet run macro applications on the first interface, we get all its declarations phase dependencies, then the interface applications, then the second interface dependencies, then the second interface applications.

Or all dependencies of from all (superclass, mixins, interfaces), and then the superclass, mixins, and interfaces in the source order?

**Comment by jakemac53** [comment]
> Oh, and the related question - the order is exactly this - first superclasses, then mixins in the source order, then interfaces in the source order, right?

Good question. I will have to think on this, cc @munificent as well.

We need to ensure we still have a well defined ordering without cycles when you go recursively down.

**Comment by jakemac53** [comment]
> Or all dependencies of from all (superclass, mixins, interfaces), and then the superclass, mixins, and interfaces in the source order?

I think this is probably the simplest answer.

**Comment by scheglov** [comment]
> I think this is probably the simplest answer.

From the implementation point of view, I'm not sure.

It is easy to think recursively - recursively ensure dependencies for the current interface, run interface, then the next interface. FWIW, this is what I currently implemented.

For non-recursive case we would need to collect all dependencies of all interfaces recursively (sic!), then iterate over them. Doable of course, but feels more complicated.

**Comment by jakemac53** [comment]
Afaik all of the options are fine from a correctness perspective, so if one makes the implementation easier I am happy to do that.

**Comment by scheglov** [comment]
My current implementation is following.
1. We process dependencies of a class declaration in the source order: superclass, then mixins, then interfaces, then the class itself.
```dart
@AddFunction('f11')
@AddFunction('f12')
class X1 extends X2 with X3 implements X4 {}

@AddFunction('f21')
@AddFunction('f22')
class X2 {}

@AddFunction('f31')
@AddFunction('f32')
mixin X3 {}

@AddFunction('f41')
@AddFunction('f42')
class X4 {}
```
results in
```dart
void f22() {}
void f21() {}
void f32() {}
void f31() {}
void f42() {}
void f41() {}
void f12() {}
void f11() {}
```

2. We process dependencies depth first.
```dart
@AddFunction('f11')
@AddFunction('f12')
class X1 implements X2, X4 {}

@AddFunction('f21')
@AddFunction('f22')
class X2 implements X3 {}

@AddFunction('f31')
@AddFunction('f32')
class X3 {}

@AddFunction('f41')
@AddFunction('f42')
class X4 {}
```
results in
```dart
void f32() {}
void f31() {}
void f22() {}
void f21() {}
void f42() {}
void f41() {}
void f12() {}
void f11() {}
```

**Comment by jakemac53** [comment]
I am curious about library cycles too - what if X2 is from an import (but part of a cycle), and X3 is from the current library?

We need X2 to run before we run on X1, but we probably don't want to affect the ordering of macros in _other libraries_, even if they are in a library cycle with the current one. Or if we do, this needs to be specifically called out I think.

**Comment by jakemac53** [comment]
Consider that many situations exist today where entire packages are one giant library cycle.

**Comment by scheglov** [comment]
Ah, thank you for reminding me. My plan was to handle cycles as well, e.g. for `class X1 implements X2, X3 {}` to process `X2` and `X3` before `X1` even if they are in different libraries. We don't have any restrictions for example to type inference - we process the whole library cycle at once. So, I see no reason to restrict macros here.

But I have not implemented the corresponding tests yet. I have to! :-)

**Comment by scheglov** [comment]
How to convert `NamedTypeAnnotation` into an `IntrospectableType`, assuming that it is one?

**Comment by scheglov** [comment]
Ah, true, we don't have yet a rule that says from which library of the library cycle we start.
Maybe sort libraries of the cycle by URI strings, in ascending order?
So, `package:test/a.dart` is processed before `package:test/b.dart`.

**Comment by jakemac53** [comment]
> Ah, true, we don't have yet a rule that says from which library of the library cycle we start.

My main concern with that approach, is that we lose the ability to do any parallelization of macros across libraries.

Today, you should mostly be able to run macros on all libraries at the same time, even if they are in a cycle. You only may have to wait at some points for a given macro to complete from a different library, if you have a super type from another library but in the current library cycle.

I believe that forcing the order in which libraries are processed for macros would forever pin us into a place where we can't run macros in parallel across libraries (and macros today do benefit from parallelism since they run on a separate thread).

**Comment by scheglov** [comment]
How about something a little more general, or maybe the same idea as you described, but more detailed?
For the declarations phase.
1. Find macro applications that don't have unsatisfied dependencies, in all libraries of the cycle, in the source order. The relative order of the libraries does not matter, we use the source order, and results go into the library specific augmentations.
2. Run all these selected applications, potentially in parallel.
3. Now more applications have dependencies satisfied, repeat from (1) until no applications to run in the declarations phase.

Yes, we should detect and break cycles in classes before the declarations phase, after the types phase, after types resolution.

**Comment by jakemac53** [comment]
https://github.com/dart-lang/language/issues/3442 is also closely related to this, and I think might further influence the design.

I don't see any way to fix that one without going with a more general approach of allowing you to attempt to introspect on any type, which may block until macros on that type have completed, and also may through if a cyclic dependency is created.

**Comment by scheglov** [comment]
I agree.
Let's consider this issue as preliminary closed, and move the discussion into #3442.
